kerastuner/engine/oracle.py

```diff
@@ -17,90 +17,241 @@
 from __future__ import division
 from __future__ import print_function
 
+import collections
 import enum
 import hashlib
+import json
 
+from tensorflow import keras
+from kerastuner.engine import hyperparameters as hp_module
+from kerastuner.engine import metrics_tracking
+from kerastuner.engine import stateful
+from kerastuner.engine import trial as trial_lib
+from ..abstractions.tensorflow import TENSORFLOW_UTILS as tf_utils
 
-class Oracle(object):
 
-    def __init__(self):
-        self.space = []
+Objective = collections.namedtuple('Objective', 'name direction')
 
-    def update_space(self, new_entries):
+
+class Oracle(stateful.Stateful):
+
+    def __init__(self,
+                 objective,
+                 max_trials,
+                 hyperparameters=None,
+                 allow_new_entries=True,
+                 tune_new_entries=True):
+        self.objective = _format_objective(objective)
+        self.max_trials = max_trials
+        if not hyperparameters:
+            if not tune_new_entries:
+                raise ValueError(
+                    'If you set `tune_new_entries=False`, you must'
+                    'specify the search space via the '
+                    '`hyperparameters` argument.')
+            if not allow_new_entries:
+                raise ValueError(
+                    'If you set `allow_new_entries=False`, you must'
+                    'specify the search space via the '
+                    '`hyperparameters` argument.')
+            self.hyperparameters = hp_module.HyperParameters()
+        else:
+            self.hyperparameters = hyperparameters
+        self.allow_new_entries = allow_new_entries
+        self.tune_new_entries = tune_new_entries
+
+        # trial_id -> Trial
+        self.trials = {}
+        # tuner_id -> Trial
```

Comment:
Can you comment/explain what the tuner_id is?
Is this the identifier for a job working on a single trial in the distributed case?

Reply:
Yep, exactly. In the multi-worker distribution strategy case we would make this the same for all workers in the same cluster